File: fbt/psitbx/psitbxlib/bspsum.c
/*
BSPSUM   computes the combination or the derivative of a combination of B-spline
         base functions.

void bspsum(V,T,nT,C,nC,X,nX,d,w)

double	*V;	the values of the combination for each X
double	*T;	the sorted knot sequence
int	nT;
double	*C;	the coefficient of each base funtion; the spline order is
		k = nT - nC > 0
int	nC;
double	*X;	the values where V must be computed; X must be in the interval
		[T(k),T(last-k+1)]
int	nX;
int	d;	the derivative order
double	*w;	work area of n k

*/

void bspsum(V,T,nT,C,nC,X,nX,d,w)

double	*V;
double	*T;
int	nT;
double	*C;
int	nC;
double	*X;
int	nX;
int	d;
double	*w;

{
 register int i, jl, jh, jn, r, s, k, kk;
 register double *ti, *tj, o, *Xi;

 k = nT - nC;
 jl = k - 1; jh = nT - k;
 for (i = 0, Xi = X; i < nX; i++, Xi++) {
  /* find j, the index in T of the larger knot <= X[i]
     note that j == nT-1 produces indexing problem */
  if (*Xi < T[jl]) jl = k - 1;
  if (*Xi > T[jh]) jh = nT - k;
  while (jh - jl > 1) {
   jn = (jl + jh) / 2;
   if (*Xi > T[jn])
    jl = jn;
   else
    jh = jn;
  }
  /* copy c in w */
  for (s=0; s < k; s++) w[s] = C[jl-s];
  /* perform derivation */
  kk = k;
  for (r = 0; r < d; r++) {
   tj = (ti = T + jl) + --kk;
   for (s = 0; s < kk; s++) {
    o = *(tj--) - *(ti--);
    w[s] = o == 0 ? 0 : (w[s] - w[s+1]) * kk / o;
   }
  }
  /* do the evaluation */
  while (kk > 1) {
   tj = (ti = T + jl) + --kk;
   for (s = 0; s < kk; s++) {
    o = *(tj--) - *ti;
    if (o != 0) o = (*Xi - *(ti--)) / o;
    w[s] = w[s] * o + (1 - o) * w[s+1];
   }
  }
  V[i] = w[0];
 }
}

File: fbt/psitbx/psitbxlib/bspsum.m
function v = bspsum(t,c,x,d,p,q)

% BSPSUM   B-spline base function combination
%   BSPSUM(T,C,X) estimates at each X the combination of spline base functions
%   defined by the knot sequence T. C contains along its first dimension the
%   coefficients of the combination, and BSPSUM returns an array of size
%   [size(X),size(C)(2:end)] or [length(X),size(C)(2:end)] if X is a vector.
%   BSPSUM(T,C,X,D) allows to estimate the D-th derivative.
%   BSPSUM(T,C,X,D,1) expects a C of size [size(C,1),size(X),DIM] and
%   returns an array of size [size(X),DIM], having applied the corresponding
%   coefficient to the corresponding X value.
%   BSPSUM(T,C,X,D,*,1) assumes distinct knots and uses a faster coding.
%   Spline order K is length(T)-size(C,1).
%   BSPSUM uses a MEX-file much faster than SPVAL in the Spline Toolbox.

% test:  plot(bspsum(-2:12,eye(12,15),linspace(0,10,101)))

if nargin < 4, d = 0; end
if nargin < 5, p = 0; end
if nargin < 6, q = all(diff(t)); end

nt = length(t);
dimc = size(c);
dimx = size(x); if length(dimx) == 2 & min(dimx) < 2, dimx = length(x); end
if nt - dimc(1) < 1
 error('Too many coefficients for specified knot sequence')
end
if p & (length(dimc)+1 < length(dimx) | ~isequal(dimc(2:length(dimx)+1),dimx))
 error('Incompatible C and X size')
end

v = bspsummex(t,c,prod(dimc(2:end)),x,prod(dimx),d,p,q);
if p
 v = reshape(v,[dimx,dimc(length(dimx)+2:end),1]);
else
 v = reshape(v,[dimx,dimc(2:end)]);
end
